### ISSUE TITLE
opt: "could not decorrelate subquery" error when using LATERAL + subquery

### ISSUE THREAD

**Issue by andy-kimball** [cockroachdb/cockroach#45856]
Repro:
```
root@:26257/defaultdb> SELECT * FROM (VALUES (1), (2)) t(x), LATERAL (VALUES (t.x+1), (t.x+2)) t2(y) WHERE (SELECT t2.y < 100);
ERROR: could not decorrelate subquery
```
EXPECTED: No error; we should be able to decorrelate that subquery.

The problem is that the `HoistJoinSubquery` rule does not match `ApplyJoin` variants. I think it could/should, and that it would fix this issue. More thought is required, though, to make sure that WHERE subquery can be hoisted.

Jira issue: CRDB-5123

**Comment by andy-kimball** [comment]
CC @RaduBerinde

**Comment by mgartner** [comment]
This was fixed by #95234. Closing.
